### ISSUE TITLE
`io_uring_prep_timeout` never fires

### ISSUE THREAD

**Issue by Oipo** [axboe/liburing#1211]
I might be misunderstanding something, but the following program never fires the timeout on my machine:

```c++
#include <liburing.h>
#include <system_error>
#include <iostream>

int main() {
    io_uring _eventQueue{};
    io_uring_params p{};
    p.flags = IORING_SETUP_DEFER_TASKRUN | IORING_SETUP_COOP_TASKRUN | IORING_SETUP_SINGLE_ISSUER;
    auto ret = io_uring_queue_init_params(2048, &_eventQueue, &p);
    if(ret < 0) {
        throw std::system_error(-ret, std::generic_category(), "io_uring_queue_init_params() failed");
    }
    ret = io_uring_register_ring_fd(&_eventQueue);
    if(ret < 0) {
        throw std::system_error(-ret, std::generic_category(), "io_uring_register_ring_fd() failed");
    }


    auto *sqe = io_uring_get_sqe(&_eventQueue);
    io_uring_prep_nop(sqe);
    io_uring_sqe_set_data(sqe, (void*)0);

    {
        sqe = io_uring_get_sqe(&_eventQueue);
        __kernel_timespec timespec{};
        timespec.tv_nsec = 500'000'000;
        io_uring_prep_timeout(sqe, &timespec, 0, 0);
        io_uring_sqe_set_data(sqe, (void *) 1);
    }

    ret = io_uring_submit(&_eventQueue);
    if(ret < 0) {
        throw std::system_error(-ret, std::generic_category(), "io_uring_submit() failed.");
    }

    int timeoutsReceived{};
    while(timeoutsReceived < 10) {
        io_uring_cqe *cqe{};
        __kernel_timespec ts{};
        ts.tv_nsec = 50'000'000;

        ret = io_uring_wait_cqes(&_eventQueue, &cqe, 1, &ts, nullptr);
        if(ret < 0) {
            throw std::system_error(-ret, std::generic_category(), "io_uring_wait_cqes() failed.");
        }
        unsigned int handled{};
        unsigned int head{};
        io_uring_for_each_cqe(&_eventQueue, head, cqe) {
            auto data = io_uring_cqe_get_data(cqe);
            if(data == (void*)1) {
                timeoutsReceived++;
                std::cout << "timeoutsReceived: " << timeoutsReceived << "\n";

                sqe = io_uring_get_sqe(&_eventQueue);
                __kernel_timespec timespec{};
                timespec.tv_nsec = 500'000'000;
                io_uring_prep_timeout(sqe, &timespec, 0, 0);
                io_uring_sqe_set_data(sqe, (void*)1);
            }
            handled++;
        }
        io_uring_cq_advance(&_eventQueue, handled);

        sqe = io_uring_get_sqe(&_eventQueue);
        io_uring_prep_nop(sqe);
        io_uring_sqe_set_data(sqe, (void*)0);

        ret = io_uring_submit(&_eventQueue);
        if(ret < 0) {
            throw std::system_error(-ret, std::generic_category(), "io_uring_submit() failed.");
        }
    }

    return 0;
}
```

Compiled with `g++ -g3 -std=c++20 -I ../external/liburing/src/include/ ../test/iouringsleeptest.cpp ../external/liburing/src/liburing.a`

kernel version: 6.9.5
liburing version: 2.7

**Comment by axboe** [comment]
This __kernel_timespec:

```
{
        sqe = io_uring_get_sqe(&_eventQueue);
        __kernel_timespec timespec{};
        timespec.tv_nsec = 500'000'000;
        io_uring_prep_timeout(sqe, &timespec, 0, 0);
        io_uring_sqe_set_data(sqe, (void *) 1);
    }

    ret = io_uring_submit(&_eventQueue);
```

goes out of scope before it gets consumed, it could be random garbage by the time the kernel sees it.

**Comment by axboe** [comment]
Feel like I should make this a more prominent note in general, as this is the second time today that someone had that. Data needs to be stable until submit returns.

**Comment by Oipo** [comment]
I have the same behaviour when ensuring that it doesn't go out of scope:

```c++
#include <liburing.h>
#include <system_error>
#include <iostream>

int main() {
    io_uring _eventQueue{};
    io_uring_params p{};
    p.flags = IORING_SETUP_DEFER_TASKRUN | IORING_SETUP_COOP_TASKRUN | IORING_SETUP_SINGLE_ISSUER;
    auto ret = io_uring_queue_init_params(2048, &_eventQueue, &p);
    if(ret < 0) {
        throw std::system_error(-ret, std::generic_category(), "io_uring_queue_init_params() failed");
    }
    ret = io_uring_register_ring_fd(&_eventQueue);
    if(ret < 0) {
        throw std::system_error(-ret, std::generic_category(), "io_uring_register_ring_fd() failed");
    }


    auto *sqe = io_uring_get_sqe(&_eventQueue);
    io_uring_prep_nop(sqe);
    io_uring_sqe_set_data(sqe, (void*)0);
    __kernel_timespec timespec{};

    sqe = io_uring_get_sqe(&_eventQueue);
    timespec.tv_nsec = 500'000'000;
    io_uring_prep_timeout(sqe, &timespec, 0, 0);
    io_uring_sqe_set_data(sqe, (void *) 1);

    ret = io_uring_submit(&_eventQueue);
    if(ret < 0) {
        throw std::system_error(-ret, std::generic_category(), "io_uring_submit() failed.");
    }

    int timeoutsReceived{};
    while(timeoutsReceived < 10) {
        io_uring_cqe *cqe{};
        timespec.tv_nsec = 50'000'000;

        ret = io_uring_wait_cqes(&_eventQueue, &cqe, 1, &timespec, nullptr);
        if(ret < 0) {
            throw std::system_error(-ret, std::generic_category(), "io_uring_wait_cqes() failed.");
        }
        unsigned int handled{};
        unsigned int head{};
        io_uring_for_each_cqe(&_eventQueue, head, cqe) {
            auto data = io_uring_cqe_get_data(cqe);
            if(data == (void*)1) {
                timeoutsReceived++;
                std::cout << "timeoutsReceived: " << timeoutsReceived << "\n";

                sqe = io_uring_get_sqe(&_eventQueue);
                timespec.tv_nsec = 500'000'000;
                io_uring_prep_timeout(sqe, &timespec, 0, 0);
                io_uring_sqe_set_data(sqe, (void*)1);
            }
            handled++;
        }
        io_uring_cq_advance(&_eventQueue, handled);

        sqe = io_uring_get_sqe(&_eventQueue);
        io_uring_prep_nop(sqe);
        io_uring_sqe_set_data(sqe, (void*)0);

        ret = io_uring_submit(&_eventQueue);
        if(ret < 0) {
            throw std::system_error(-ret, std::generic_category(), "io_uring_submit() failed.");
        }
    }

    return 0;
}
```

**Comment by Oipo** [comment]
> Feel like I should make this a more prominent note in general, as this is the second time today that someone had that. Data needs to be stable until submit returns.

It would be nice if, f.e., compiling liburing with the address sanitizer would check all to-be-submitted SQE's for memory access in `__io_uring_submit`. That way, a tool would check for out-of-scope-ness instead of having people not read properly.

**Comment by axboe** [comment]
I'll take a look at your test case.

> It would be nice if, f.e., compiling liburing with the address sanitizer would check all to-be-submitted SQE's for memory access in __io_uring_submit. That way, a tool would check for out-of-scope-ness instead of having people not read properly.

Not sure how feasible that is, as the memory pointed to is opaque at that point. It would need to understand the command types and explicitly check for the memory passed in.

**Comment by axboe** [comment]
I think I see what it is - you keep posting NOPs in there, which means your wait side will never really wait on events as some are always available. This means that the kernel is never entered, and hence that the timeout task_work is never run. For DEFER_TASKRUN, you'll want a call to ```io_uring_get_events()``` to process these, eg before your ```io_uring_for_each_cqe()``` loop. If you add that, I suspect it should work just fine.

The test is a bit odd in that way, as any NOP will immediately complete and hence you're really just busy looping around submitting a nop and processing its completion, with the occasional timeout being processed too.

**Comment by Oipo** [comment]
Ah, yes. That was exactly the issue. Thanks! I'll close this issue now. I'm confident I can get something working with the address sanitizer, I'll open a PR when I've got some free time for that and we'll see if it is something that could be maintained long term or if I'll just use it for myself.

Regarding the test, yes, I reduced the issue to the least amount of lines I could. But I use NOPs to post non-io_uring related events either from the current thread or from another thread through prep_msg_ring.

**Comment by axboe** [comment]
> 'm confident I can get something working with the address sanitizer, I'll open a PR when I've got some free time for that and we'll see if it is something that could be maintained long term or if I'll just use it for myself.

That'd be great! Would be a lot easier to tell folks to build liburing with sanitizer and have them verify it doesn't trigger for an issue like this, or just as a sanity thing in general to verify that you actually did it right. These kinds of issues tend to be somewhat unpredictable in nature, obviously.
